### PR TITLE
dockertools: fix buildLayeredImage nix-store permissions

### PR DESCRIPTION
**Pull request by johanot**
###### Motivation for this change
Fix bug in `dockerTools.buildLayeredImage` that causes non-root containers to be not runnable.

Pardon my python, but I was unsure how - in the prettiest way - to hook into the existing logic around /nix/store dir-creation.

example docker error message:
```
docker: Error response from daemon: OCI runtime create failed: container_linux.go:349: starting container process caused "exec: \"/nix/store/3avafdzgh694z5m3rgw2vvlgsfz2gmkl-coredns-1.6.6/bin/coredns\": stat /nix/store/3avafdzgh694z5m3rgw2vvlgsfz2gmkl-coredns-1.6.6/bin/coredns: permission denied": unknown.
```

The cause of the error is that `/nix` and `/nix/store` has wrong permissions inside the final image, i.e.:
```
bash-4.4# ls -dl /nix/store/
drw-r--r-- 32 root root 34 Jan  1  1970 /nix/store/
```
The top-level directories lack world execute permissions, which, is ok as long we run the image as root, but fails when the image is run as an unprivileged user.

###### Steps to reproduce

1. cherry-pick 4fc255e from my branch onto master
2. `docker load <$(nix-build . -A dockerTools.examples.bashLayeredWithUser)`
3. `docker run --rm -it -u somebody bash-layered-with-user /bin/echo "it works"`

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @LnL7 @utdemir

### CODE REVIEW

File: pkgs/build-support/docker/examples.nix [hunk 1]
```diff
@@ -382,4 +382,40 @@ rec {
     contents = pkgs.bashInteractive;
   };
 
+  # buildLayeredImage with unprivileged user
+  bashLayeredWithUser =
+  let
+    nonRootShadowSetup = { user, uid, gid ? uid }: with pkgs; [
```
Comment (utdemir) [inline pkgs/build-support/docker/examples.nix:388]: Do we need to set up `/etc/shadow` to spin up the container? AFAIK it should work without until something tries to lookup the user information, which is not the case in this test case. 

It should make the test case a lot smaller to remove `nonRootShadowSetup`.

**Review by utdemir — COMMENTED** [review]
Thank you for the change! 

A similar PR was submitted a few days ago too: https://github.com/NixOS/nixpkgs/pull/93811, but we were waiting for a test. Since this one has also tests, @adrian-gierakowski do you mind if we continue from this PR? The changes are almost identical save the name of the function and this one has a test case.

**Comment by adrian-gierakowski** [comment]
@utdemir sounds good to me

File: pkgs/build-support/docker/examples.nix (hunk 1 above)
Comment (adrian-gierakowski) [reply pkgs/build-support/docker/examples.nix:388]: As you said it’s not strictly necessary but I think it’s useful to have an example of how to do it and ideally with a comment explaining when one might need it.

**Review by utdemir — APPROVED** [review]
The tests have passed on my system, I think this PR is good to merge.

File: nixos/tests/docker-tools.nix [hunk 2]
```diff
@@ -79,6 +79,16 @@ import ./make-test-python.nix ({ pkgs, ... }: {
             "docker rmi ${examples.nix.imageName}",
         )
 
+    with subtest(
+        "Ensure (layered) nix store has correct permissions "
+        "and that starting un-privileged containers works"
```
Comment (roberth) [inline nixos/tests/docker-tools.nix:84]:
```suggestion
        "and that the container starts when its process does not have uid 0"
```

"Priviliged" refers to containers that run with capabilities that could harm the host.

File: pkgs/build-support/docker/examples.nix [hunk 3]
```diff
@@ -382,4 +382,40 @@ rec {
     contents = pkgs.bashInteractive;
   };
 
+  # buildLayeredImage with unprivileged user
```
Comment (roberth) [inline pkgs/build-support/docker/examples.nix:385]:
```suggestion
  # buildLayeredImage with non-root user
```

**Review by roberth — COMMENTED** [review]
Just one little detail: "priviliged" refers to containers that run with extra capabilities that could harm the host.
Otherwise looks good. Thank you for adding a test case!

**Comment by johanot** [comment]
@roberth Valid point! I applied your suggestion just now using the fancy github feature. :) I'll squash those extra commits into my test-commit in a sec.

**Comment by srhb** [comment]
@ofborg test docker-tools

**Comment by srhb** [comment]
Ofborg is happy too! Thanks :)

**Pull request merged**